Sources/NIOConcurrencyHelpers/lock.swift

```diff
@@ -45,7 +47,7 @@ public final class Lock {
     public init() {
 #if os(Windows)
         InitializeSRWLock(self.mutex)
-#else
+#elseif !os(WASI)
```

Comment:
I don't think we can merge a PR that silently disables these primitives on WASI without being _extremely_ confident that we'll never need them. Is there any way we can replace this with a feature-detection on threading?

Reply:
Swift `main` snapshots support `#if _runtime(_multithreaded)` checks, which I'm now using here.